block/pending_blocks.go

```diff
@@ -1,44 +1,55 @@
 package block
 
 import (
+	"sort"
 	"sync"
 
 	"github.com/rollkit/rollkit/types"
 )
 
 // PendingBlocks maintains blocks that need to be published to DA layer
 type PendingBlocks struct {
-	pendingBlocks []*types.Block
+	pendingBlocks map[uint64]*types.Block
 	mtx           *sync.RWMutex
 }
 
 // NewPendingBlocks returns a new PendingBlocks struct
 func NewPendingBlocks() *PendingBlocks {
 	return &PendingBlocks{
-		pendingBlocks: make([]*types.Block, 0),
+		pendingBlocks: make(map[uint64]*types.Block),
 		mtx:           new(sync.RWMutex),
 	}
 }
 
 func (pb *PendingBlocks) getPendingBlocks() []*types.Block {
 	pb.mtx.RLock()
 	defer pb.mtx.RUnlock()
-	return pb.pendingBlocks
+	blocks := make([]*types.Block, 0)
```

Comment:
To avoid re-allocation of memory we should use the length of pb.pendingBlocks since it is known. 
```suggestion
	blocks := make([]*types.Block, 0, len(pb.pendingBlocks))
```